Framework/API/inc/MantidAPI/Algorithm.h

```diff
@@ -306,6 +307,34 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
   /// Virtual method - must be overridden by concrete algorithm
   virtual void exec() = 0;
 
+  virtual void execDistributed() { exec(); }
+  virtual void execMasterOnly();
+  virtual void execNonMaster(){};
+
+  /** Get correct execution mode based on input storage modes for an MPI run.
+   *
+   * The default implementation returns ExecutionMode::Invalid. Classes
+   * inheriting from Algorithm can re-implement this if they support execution
+   * with multiple MPI ranks. May not return ExecutionMode::Serial, because that
+   * is not a "parallel" execution mode. */
+  virtual MPI::ExecutionMode getParallelExecutionMode(
+      const std::map<std::string, MPI::StorageMode> &storageModes) const;
+
+  /** Get storage mode for an output workspace.
+   *
+   * The default implementation returns the storage mode of the input workspace.
+   * If there is more than one input workspace it throws an exception. Classes
+   * inheriting from Algorithm can re-implement this when needed. The workspace
+   * is identified by the name if its property in the Algorithm (*not* the name
+   * of the Workspace). */
+  virtual MPI::StorageMode
+  getStorageModeForOutputWorkspace(const std::string &propertyName) const;
```

Comment:
Yes, `propagateWorkspaceStorageMode()` will call `getStorageModeForOutputWorkspace` for each output.